### PR TITLE
Use new instance of HDFS object to fetch NameNode token as URI is unique

### PR DESCRIPTION
**Pull request by djaiswal83**
Use new instance of HDFS object to fetch NameNode token as URI is unique with updated Kerberos Principal.

With older logic if there exists a FileSystem cache, the cache would blow up very quickly.
Earlier, the UGI for namenode used to be just the proxy user name. If a job with same user requested FileSytem object, it would be fetched from the cache.
With update to Kerberos principal to <user name>/az_<host name>_<exec_id>, the UGI became unique to an execution thus blowing the FileSystem cache up.
Morever, the entry is never fetched from FileSystem cache as it is unique to that particular execution rendering the cache entry useless.
This PR alleviates this by creating a new instance of FS and closing it in the same method.

EDIT : Issac okayed the current logic.

### CODE REVIEW

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/security/HadoopSecurityManager_H_2_0.java [hunk 1]
```diff
@@ -567,44 +567,68 @@ private void fetchJobTrackerToken(final String userToProxy, final Props props,
...
-      // check if we get the correct FS, and most importantly, the conf
-      logger.info("Getting DFS token from " + fs.getUri());
-      final Token<?>[] fsTokens =
-          fs.addDelegationTokens(getMRTokenRenewerInternal(new JobConf()).toString(), cred);
-      if (fsTokens.length == 0) {
-        logger.error("Failed to fetch DFS token for ");
-        throw new HadoopSecurityManagerException(
-            "Failed to fetch DFS token for " + userToProxy);
-      }
+      //final FileSystem fs = FileSystem.get(HadoopSecurityManager_H_2_0.this.conf);
```
Comment (arvindpruthi) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/security/HadoopSecurityManager_H_2_0.java:572]: Commented out code should be removed.

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/security/HadoopSecurityManager_H_2_0.java [hunk 2]
```diff
@@ -567,44 +567,68 @@ private void fetchJobTrackerToken(final String userToProxy, final Props props,
...
 
       // getting additional name nodes tokens
       final String otherNamenodes = props.get(
-          HadoopSecurityManager_H_2_0.OTHER_NAMENODES_TO_GET_TOKEN);
+              HadoopSecurityManager_H_2_0.OTHER_NAMENODES_TO_GET_TOKEN);
       if ((otherNamenodes != null) && (otherNamenodes.length() > 0)) {
         logger.info("Fetching token(s) for other namenode(s): " + otherNamenodes);
         final String[] nameNodeArr = otherNamenodes.split(",");
-        final Path[] ps = new Path[nameNodeArr.length];
+        /*final Path[] ps = new Path[nameNodeArr.length];
```
Comment (arvindpruthi) [inline azkaban-hadoop-security-plugin/src/main/java/azkaban/security/HadoopSecurityManager_H_2_0.java:586]: Where is the "/* closing?

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/security/HadoopSecurityManager_H_2_0.java [hunk 3]
```diff
@@ -567,40 +567,58 @@ private void fetchJobTrackerToken(final String userToProxy, final Props props,
...
+                                          final String userToProxy, final URI uri)
+          throws IOException, HadoopSecurityManagerException {
+    FileSystem fs = null;
+    try {
+      if (uri == null) {
+        fs = FileSystem.newInstance(conf);
       } else {
-        logger.info(
-            HadoopSecurityManager_H_2_0.OTHER_NAMENODES_TO_GET_TOKEN + " was not configured");
+        fs = FileSystem.newInstance(uri, conf);
```
[the remark this answers is not in the document]
Comment (djaiswal83) [reply azkaban-hadoop-security-plugin/src/main/java/azkaban/security/HadoopSecurityManager_H_2_0.java:602]: For other name nodes the URI of those name nodes needs to be provided explicitly.

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/security/HadoopSecurityManager_H_2_0.java [hunk 4]
```diff
@@ -567,40 +567,58 @@ private void fetchJobTrackerToken(final String userToProxy, final Props props,
...
+      try {
+        final Token<?>[] fsTokens = fs.addDelegationTokens(renewer, cred);
+        for (int i = 0; i < fsTokens.length; i++) {
+          final Token<?> fsToken = fsTokens[i];
+          logger.info(String.format(
+                  "DFS token from namenode pre-fetched, token kind: %s, token service: %s",
+                  fsToken.getKind(), fsToken.getService()));
+        }
+      } catch (Exception e) {
+        logger.error("Failed to fetch DFS token for ");
```
[the remark this answers is not in the document]
Comment (djaiswal83) [reply azkaban-hadoop-security-plugin/src/main/java/azkaban/security/HadoopSecurityManager_H_2_0.java:615]: Thanks. It was like that forever, will fix it.

File: azkaban-hadoop-security-plugin/src/main/java/azkaban/security/HadoopSecurityManager_H_2_0.java [hunk 5]
```diff
@@ -567,40 +567,58 @@ private void fetchJobTrackerToken(final String userToProxy, final Props props,
...
+    } else {
+      logger.info(
+              HadoopSecurityManager_H_2_0.OTHER_NAMENODES_TO_GET_TOKEN + " was not configured");
+    }
+  }
+
+  private void fetchNameNodeTokenInternal(final String renewer, final Credentials cred,
+                                          final String userToProxy, final URI uri)
+          throws IOException, HadoopSecurityManagerException {
+    FileSystem fs = null;
```
[the remark this answers is not in the document]
Comment (djaiswal83) [reply azkaban-hadoop-security-plugin/src/main/java/azkaban/security/HadoopSecurityManager_H_2_0.java:597]: Thanks, will do.

**Pull request merged**